### PR TITLE
[bloom] convert script tweaks

### PR DESCRIPTION
**Pull request by stas00**
@younesbelkada, could you please have a look

1. when creating a small model for testing the assert for unexpected keys breaks the conversion - I think we should either not assert or perhaps warn instead?

2. also don't try to set up dtype if it's `None`

actually for the latter if `torch_dtype` is not defined, shouldn't we try to derive the dtype from the Meg-DS checkpoint dtype? since currently it'll just create fp32 weights, ignoring the actual dtype of the weights.

Thank you

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Hi @stas00 !

Thanks for spotting this! Regarding your points:
1- I am just curious why the missing keys test fails for small models? Could you tell me which keys are missing? In my experience this test was always passing but I am fine to put a warning instead. 
2- Totally fine with that! 

Thanks a lot 💪

**Comment by stas00** [comment]
sorry, forgot to follow up here.

It was just a small model that I created on the fly.

that assert is not user-friendly as it just fails w/o telling which keys are unexpected. if it were to tell which keys are unexpected I would be able to answer your question

**Comment by stas00** [comment]
@younesbelkada, this is still an issue - we are training a few variations of bloom for m4 https://github.com/huggingface/m4/blob/text_pretraining/experiments/pretraining/text_pretraining/narrow_gpt.slurm and the conversion fails in 2 asserts:

```
  File "src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py", line 194, in convert_bloom_checkpoint_to_pytorch
    assert not other_keys.unexpected_keys
AssertionError
```
and if the above removed, then next in:
```
  File "src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py", line 200, in convert_bloom_checkpoint_to_pytorch
    assert not missing_keys
AssertionError
```

it's fine then at converting. so besides my PR a 2nd assert is an issue as well.

Thank you!

cc for awareness: @TevenLeScao

File: src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py [hunk 1]
```diff
@@ -191,7 +191,7 @@ def convert_bloom_checkpoint_to_pytorch(
                     tensors[key] = tensors[key] / pretraining_tp
 
             other_keys = model.load_state_dict(tensors, strict=False)
-            assert not other_keys.unexpected_keys
+            # assert not other_keys.unexpected_keys
```
Comment (younesbelkada) [inline src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py:194]:
```suggestion
            assert not other_keys.unexpected_keys, f"The keys {other_keys.unexpected_keys} are unexpected"
```

**Review by younesbelkada — COMMENTED** [review]
Thanks @stas00 for the heads up! 
IMO it is a bit surprising that the first `assert` does not pass, could you share with me your config file as I don't have access to JZ anymore? 🙏 Also if I can have access to the weights (or some dummy weights) so that I can try them out as well. I believe the fix should be very easy, and sorry for the delay!
Alternativement I think that you can commit my suggestion and tell me which key is missing

**Comment by stas00** [comment]
grr, my apologies, @younesbelkada - it proved to be a misinformation - the pretrained model was meg-ds gpt2 model and not bloom :( sorry about that.

but yes, your suggestion of printing out the unexpected keys rather than how it was before sounds great to me

Let's do that for both asserts?

**Comment by stas00** [comment]
for posterity I ended up using this Megatron-Deepspeed conversion script 
https://github.com/bigscience-workshop/bigscience/tree/aa872e754106f6678e8a9dac8c6962404ba39a6d/train/tr1-13B-base#checkpoint-conversion-and-upload
we wrote when developing [pre-bloom tr1-13b-en model](https://github.com/bigscience-workshop/bigscience/tree/aa872e754106f6678e8a9dac8c6962404ba39a6d/train/tr1-13B-base) and then using HF's GPT2 modeling code to generate with it. It's not perfect as GPT2 != gpt2 in Meg-DS - 3 differences are https://github.com/bigscience-workshop/Megatron-DeepSpeed/issues/138 but it more or less works.

**Comment by younesbelkada** [comment]
Ahh I see thanks a lot for the clarification! 
Yes I think we should update the asserts ;) ! Also I think it might be useful to have a script to convert meg-ds gpt2 models to HF format, where do you think we should put this file? Or maybe just adding an arg `convert-gpt2` on the current file would work too, your call!

**Comment by stas00** [comment]
I updated the PR to improve the 2nd assert, so I think it's good to merge now.

> Also I think it might be useful to have a script to convert meg-ds gpt2 models to HF format, where do you think we should put this file?

I'm not sure since it depends on `Megatron-Deepsped`'s internal files: https://github.com/bigscience-workshop/Megatron-DeepSpeed/tree/main/tools/convert_checkpoint and it's specific to this fork of Meg-DS (or rather nobody maintains the parent fork that is under MSFT)

**Review by younesbelkada — APPROVED** [review]
LGTM! Thanks a lot for correctly taking care of the `torch_dtype` and having a more user-friendly conversion script!

regarding your suggestion for meg-ds GPT2, yes I guess that would work too!

**Comment by stas00** [comment]
There is also https://github.com/huggingface/transformers/tree/main/src/transformers/models/megatron_gpt2 but that one can only handle Megatron-LM generated checkpoints (not Megatron-Deepspeed ones).

**Comment by younesbelkada** [comment]
> There is also https://github.com/huggingface/transformers/tree/main/src/transformers/models/megatron_gpt2 but that one can only handle Megatron-LM generated checkpoints (not Megatron-Deepspeed ones).
That one folder would fit best I think, happy to open a PR there, let me know!

**Comment by stas00** [comment]
I pushed the new doc here. 

Do you think we want a separate PR for that? can move it there if preferable. in a way it summarizes all the discussions of this PR.

whatever works.

**Comment by younesbelkada** [comment]
this is perfectly fine for me thanks a lot @stas00 !
Gently pinging @sgugger for a final review / approval 
Thank you!

**Review by sgugger — APPROVED** [review]
Thanks a lot!

**Pull request merged**